lib/view_model.rb

```diff
@@ -232,39 +232,100 @@ def has_access_check_error?
     @access_check_error.present?
   end
 
+  # Check that the user is permitted to view the record in its current state, in
+  # the given context. To be overridden by viewmodel implementation.
   def visible?(context: self.class.new_serialize_context)
     true
   end
 
+  # Check that the record is eligible to be changed in its current state, in the
+  # given context. During deserialization, this is called before any edits have
+  # taken place, and if false, an error will be raised if an edit is attempted.
+  # To be overridden by viewmodel implementations.
+  def editable?(deserialize_context: self.class.new_deserialize_context)
+    visible?(context: deserialize_context)
+  end
+
+  # Check that the attempted changes to this record are permitted in the given
+  # context. To be overridden by viewmodel implementations.
+  def valid_edit?(deserialize_context: self.class.new_deserialize_context, changes:)
+    true
+  end
+
+  # During deserialization, returns true if the viewmodel was `editable?` before
+  # any changes were attempted. Takes an error_view on which to optionally set
+  # the access check error, which can be used when delegating to a parent view's
+  # editability.
+  def was_editable?(error_view:)
+    unless @initial_editable_state
```

Comment:
viewmodel implementations don't share an initialize, so it can't be quite as clean as you wanted. But I'll prepare the error and clear after was_editable!.